subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -147,65 +162,73 @@ public void handleLostState() {
             LOGGER.warn("Watching not supported, not tracking changes between builds: {}", ex.getMessage());
-            stopWatching();
+            return stopWatching(currentRoot);
         } catch (Exception ex) {
             LOGGER.error("Couldn't update watches, not watching anymore", ex);
-            stopWatching();
+            return stopWatching(currentRoot);
         }
     }
 
     /**
      * Stop watching the known areas of the file system, and invalidate
-     * the parts that have been changed since calling {@link #startWatching()}}.
+     * the parts that have been changed since calling {@link #startWatching(SnapshotHierarchy)}}.
      */
     private void stopWatching() {
-        updateWatchRegistry(fileWatcherRegistry -> {
+        getRoot().update(this::stopWatching);
+    }
+
+    private SnapshotHierarchy stopWatching(SnapshotHierarchy currentRoot) {
+        if (watchRegistry != null) {
             try {
+                FileWatcherRegistry toBeClosed = watchRegistry;
                 watchRegistry = null;
-                delegatingUpdateFunctionDecorator.setSnapshotDiffListener(null);
-                fileWatcherRegistry.close();
+                delegatingUpdateFunctionDecorator.stopListening();
+                toBeClosed.close();
             } catch (IOException ex) {
                 LOGGER.error("Couldn't fetch file changes, dropping VFS state", ex);
-                getRoot().update(SnapshotHierarchy::empty);
             }
-        });
-        getRoot().update(SnapshotHierarchy::empty);
+        }
+        return currentRoot.empty();
     }
 
-    private void handleWatcherRegistryEvents(String eventsFor) {
-        updateWatchRegistry(watchRegistry -> {
-            FileWatcherRegistry.FileWatchingStatistics statistics = watchRegistry.getAndResetStatistics();
-            LOGGER.warn("Received {} file system events {}", statistics.getNumberOfReceivedEvents(), eventsFor);
-            if (statistics.isUnknownEventEncountered()) {
-                LOGGER.warn("Dropped VFS state due to lost state");
-                stopWatching();
-            }
-            if (statistics.getErrorWhileReceivingFileChanges().isPresent()) {
-                LOGGER.warn("Dropped VFS state due to error while receiving file changes", statistics.getErrorWhileReceivingFileChanges().get());
-                stopWatching();
-            }
-        });
+    private SnapshotHierarchy handleWatcherRegistryEvents(SnapshotHierarchy currentRoot, String eventsFor) {
+        if (watchRegistry == null) {
+            return currentRoot.empty();
+        }
+        FileWatcherRegistry.FileWatchingStatistics statistics = watchRegistry.getAndResetStatistics();
+        LOGGER.warn("Received {} file system events {}", statistics.getNumberOfReceivedEvents(), eventsFor);
+        if (statistics.isUnknownEventEncountered()) {
+            LOGGER.warn("Dropped VFS state due to lost state");
+            return stopWatching(currentRoot);
```

Comment:
Here `stopWatching()` also feels weird. This method does more than that, maybe `stopWatchingAndInvalidate()`?